### DISCUSSION TITLE
Why milvus use Kafka/Pulsar for querying data?

### DISCUSSION THREAD

**Discussion by ducanh997** [milvus-io/milvus#31525] [Q&A and General discussion] [answered]
My question is why does Milvus use Kafka/Pulsar for read queries instead of directly calling the query node? I'm using Kafka, Etcd, and Minio on physical servers and other services in Kubernetes, so I believe this architecture will add extra latency to read queries.

**Comment by yhmo** [comment]
Search requests are passed directly from proxy to query nodes, only insert/delete/upsert requests are stored into kafka/pulsar as write-ahead-log.

**Reply by ducanh997** [reply]
But when I shut down Pulsar, all search requests are pending forever.

**Reply by yhmo** [reply, marked as answer]
Because milvus uses a message queue to maintain data consistency.
Milvus data consistency is based on timetick synchronization:
 https://milvus.io/docs/v2.3.x/time_sync.md

**Reply by ducanh997** [reply]
Thank you. I've changed the consistency to "Eventually," so now I can execute queries even if the message queue is shut down.
